### PR TITLE
Add more filters to the mod download dialog

### LINKED ISSUE #588: [Feature Request] Filter by Modrinth category when using the mod downloader
### Role

I play modded minecraft

### Suggestion

Add Options to filter mods by category under the filter button in the download

### Benefit

Make finding mods for specific needs easier, i.e. looking for just optimization mods

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Add Options to filter mods by category under the filter button in the download mods window,
these options would be individually toggleable to you can filter for more than one category at once,
If the Modrinth API allows pulling the list of available categories it would be nice for these filters to dynamically update if Modrinth makes changes to their categories, i.e. if a new category is added it will automatically add a filter in PrisimLauncher

Modrinth's API allows searching with categories via facets see: https://docs.modrinth.com/docs/tutorials/api_search/

I've made a mock-up in photoshop as to what this could look like:
MockUp

### LINKED ISSUE #962: New column in mods list to indicate mod environments
### Role

I play modded Minecraft using modpacks

### Suggestion

I would like a new column in the mods list that specifies whether each mod is client-only or client/server

### Benefit

This would allow me to more easily discover my client-only mods that can safely be updated without the instance becoming incompatible with servers

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

- I play on a modpack server which means that I must ensure that my mod versions stay compatible with those on the server.
- Client-side only mods will never have a server-side restriction on their version since they are not on the server.
    - As such, client-side mods can be updated freely without worrying about server-side compatibility.
- I would like a new column in the mods list that tells me which mods are client-side only so that I can update them.
- I would also be happy with a much more automated feature that can restrict updates to mods based on a server's mod versions but that sounds like it would be considerably more work.

### LINKED ISSUE #1042: Search all versions by default
### Role

Update and find mods after a new MC version releases

### Suggestion

Replace the 'search', 'strict', 'major', 'any' by a presentation of the results that shows if the found mod is compatible (strict), maybe compatible (major), probably not compatible (other version)

### Benefit

Sometimes, when searching from a mod, it doesn't get found because it is tagged incorrectly (e.g., on Modrinth), having a result that informs about how compatible the found item is better than no result at all

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

## Before: 

Search ____________________bleh
* [x] 1.19.4
* [ ] 1.19
* [ ] any

**not found**

## After: 

Search __________________bleh

* bleh version x.y.z (modrinth) for minecraft 1.19.3 (not an exact version match __color background in yellow__)
* blehonasm x.y.z (curseforge) for minecraft 1.18 (probably incompatible __color background in red__)

### LINKED ISSUE #1260: Add a filter to "Download mods" screen for already installed mods
### Role

I play minecraft on custom modpacks

### Suggestion

In the "Download mods" screen add a filter checkbox to hide installed mods

### Benefit

It would make it easier to search for new mods, specially in search by popular

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Should be off by default, easily toggled in the "Filter Options" section alongside the version match options.
Should be easy to implement since installed mods are already highlighted on the list.

### LINKED ISSUE #1314: Ability to sort download search by category
### Role

For downloading mods

### Suggestion

I would like to be able to sort the download search screen by mod category.

### Benefit

Because it helps me to find the mods I need, and having to search for them using my browser then find them in Prism Launcher's download search is very annoying.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Other features for narrowing down search results in the download search would be awesome, but categories are the main thing that I need. I often sort by categories to find mods, but after I had some trouble trying to put manually downloaded mods into my modpack, people on the Prism discord said that it was better to download mods from inside the launcher due to manual mods having many issues. However, finding mods using a browser search tool, then finding them again in Prism Launcher is not a workable solution for me in the long term, due to my ADHD.
Thank you in advance for considering my suggestion.

### LINKED ISSUE #1568: filter client side
### Role

when downloading from modrinth i would like to filter 

### Suggestion

edit > mods > download mods

### Benefit

this would be nice 

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

_No response_

### LINKED ISSUE #1616: Add support for downloading mods for a different loader
### Role

If I want to use a mod such as Fabricated Forge, I cannot download forge mods in the built-in downloader.

### Suggestion

Add an option to ignore loader in the mod downloader/detect when such mods (i.e. Fabricated Forge) are present.

### Benefit

I can use the built in downloader to download mods for other loaders.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Afaik there aren't any issues that are the same as this/similar, but I have not looked much.

### LINKED ISSUE #2102: Download mods filter options doesn't work without clearing the text field
### Operating System

macOS

### Version of Prism Launcher

8.0

### Version of Qt

6.6.0

### Description of bug

Download mods filter options doesn't work without clearing the text field, once I change the filter options and click search it doesn't search once again, we need to clear the text field (of search query) and type it once again

https://github.com/PrismLauncher/PrismLauncher/assets/73608287/daafe3cd-b426-41ea-a54c-dd40ab0b59f0

### Steps to reproduce

1. Create a new instance with the latest version of minecraft and a mod loader (fabric or forge)
2.  Edit instance and go to Mods, then click download mods
3. Search for a mod that is not compatible with the current version of minecraft (outdated)
4. Change the filter options to any version and click search, you will notice that the filter options doesn't take affect

A quick way to fix it is to clear the text field and type it again, it will work

### Suspected cause

I'm unfamiliar with this repo and QT in general

Simple fix: to fix this in code, the click search button should check if the filter options has changed, or re search anyway even if the search query didn't changed

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Trial97**
This PR adds filters for the mod download dialog.
As a side effect, there were added new columns in the mod folder page( side, loaders and mcVersions).
ToDo:
- [x] display to user extra mod information:
    - [x] environment: if the mod is client side or server of both fixes #962
    - [x] supported loaders
    - [x] supported Minecraft versions
    - [x] release type
- [x] add filter to the GUI: 
    - [x] revamp the Minecraft version select
    - [x] add loaders filter
    - [x] add client-side filter
    - [x] add filter to hide installed mods  
    - [x] add release type filter
    - [x] categories? (this is I'm unsure if is gonna be done in this PR or not)
- [x] connect the GUI with actual filters:
    - [x] Minecraft version fixes #1042
    - [x] loaders fixes #1616
    - [x] client-side fixes #1568
    - [x] hide installed mods fixes #1260
    - [x] release type
    - [x] categories fixes #1314 , fixes  #588
- [x] disable dependency search for incompatible mods(if loaders do not match or the Minecraft version doesn't match)

### CODE REVIEW

**Comment by TheKodeToad** [comment]
maybe the filters should be moved to the left

**Comment by Trial97** [comment]
Get me a mockup and I will try to implement it.

**Comment by TheKodeToad** [comment]
I think I already implemented that when I was attempting this before (I didn't get anywhere, don't worry), but I had a panel to the left of the mods list.
I think that's probably more efficient in terms of space... or you could simply have a tab for each option.

**Comment by Trial97** [comment]
Right now I did go with tabs for each option.
Also regarding minecraft version selection: do you know any good implementation of a multiselect-seachable-combobox for qt?  Or a better way to select multiple Minecraft versions? (Right now it selects multiple but is ugly and I can't search for a specific version in the combobox)
I wanted to try and copy the modrinth approach for the version select but failed because of qt.
Also, can you link the branch where you attempted the implementation?(I would love to steal some code)

**Comment by TheKodeToad** [comment]
I don't think I got any further than UI

**Comment by Trial97** [comment]
> I don't think I got any further than UI

Right now I don't need it anymore. 
Just a small note I'm bad at UI, so if you see anything that can be improved let me know(and how to do it).

**Comment by Trial97** [comment]
Should I add a warning when downloading mods that are not meant for the current Minecraft version or loader?
And if yes should it be just a small text near the mentioned filters or on the review mods page?

**Comment by TheKodeToad** [comment]
Oops, I assigned myself by mistake.
I guess that can be done in another PR?

**Comment by Trial97** [comment]
I just want to polish a little this big PR.
It doesn't need to be in this PR but I would prefer if it is.
Still, the question remains on how should I notify the user about that.

File: launcher/minecraft/mod/ModFolderModel.cpp [hunk 1]
```diff
@@ -62,12 +65,16 @@
...
+    m_column_names = QStringList(
+        { "Enable", "Image", "Name", "Version", "Last Modified", "Provider", "Side", "Loaders", "Miecraft Versions", "Release Type" });
+    m_column_names_translated = QStringList({ tr("Enable"), tr("Image"), tr("Name"), tr("Version"), tr("Last Modified"), tr("Provider"),
+                                              tr("Side"), tr("Loaders"), tr("Miecraft Versions"), tr("Release Type") });
+    m_column_sort_keys = { SortType::ENABLED,  SortType::NAME, SortType::NAME,    SortType::VERSION,     SortType::DATE,
+                           SortType::PROVIDER, SortType::SIDE, SortType::LOADERS, SortType::MC_VERSIONS, SortType::RELEASE_TYPE };
+    m_column_resize_modes = { QHeaderView::Interactive, QHeaderView::Interactive, QHeaderView::Stretch,     QHeaderView::Interactive,
+                              QHeaderView::Interactive, QHeaderView::Interactive, QHeaderView::Interactive, QHeaderView::Interactive,
+                              QHeaderView::Interactive, QHeaderView::Interactive };
+    m_columnsHideable = { false, true, false, true, true, true, true, true, true, true };
```
Comment (TheKodeToad) [inline launcher/minecraft/mod/ModFolderModel.cpp]:
```suggestion
    m_columnsHideable = { false, true, false, true, true, true, false, false, false, false };
```
although it doesn't seem that useful in the first place

Comment (Trial97) [reply launcher/minecraft/mod/ModFolderModel.cpp]: Just let the user hide the columns so, I will not accept this change.

Comment (TheKodeToad) [reply launcher/minecraft/mod/ModFolderModel.cpp]: oops

**Comment by castilio20** [comment]
it seems that cf filter doesnt use the mc version that you are it use the latest as deffault (aka your on 20.1 and the filder default would open on latest 20.4)

**Comment by Trial97** [comment]
> it seems that cf filter doesnt use the mc version that you are it use the latest as deffault (aka your on 20.1 and the filder default would open on latest 20.4)

Thanks! This should be fixed now.

**Comment by TheKodeToad** [comment]
I really don't think the new columns are too useful. Why not add more info to the bottom pane?

**Comment by Trial97** [comment]
Because here I want to sort and filter the mod list after them: if I do add them at the bottom of the page I would need to go through each mod to determine if for example is for client or server side, or if it is updated to the latest minecraft version. 
But I don't understand what exactly the inconvenience of having them as columns: you as a user can control what columns are displayed, so if you feel that they do not get you any information just hide the column and you are done.

**Comment by pizzav-xyz** [comment]
When will this be added?

**Comment by Trial97** [comment]
In 9.0 hopefully, but the UI is not the best and TheKodeToad said she will look to see how can be improved

File: launcher/ui/widgets/ModFilterWidget.cpp [hunk 2]
```diff
@@ -115,6 +117,13 @@ ModFilterWidget::ModFilterWidget(MinecraftInstance* instance, bool extended, QWi
     prepareBasicFilter();
 }
 
+bool ModFilterWidget::eventFilter(QObject* obj, QEvent* evt)
+{
+    if ((obj != ui->versions && obj != ui->version) || evt->type() != QEvent::Wheel)
+        return QTabWidget::eventFilter(obj, evt);
+    return true;
+}
```
Comment (TheKodeToad) [inline launcher/ui/widgets/ModFilterWidget.cpp]: Couldn't this be in CheckComboBox::eventFilter

Comment (Trial97) [reply launcher/ui/widgets/ModFilterWidget.cpp]: will still need this function for `ui->version` as that one is the normal checkbox.
I can move a part of it but I will still need to keep the eventFilter on `ModFilterWidget` just for `ui->version`, are you ok with that?

Comment (TheKodeToad) [reply launcher/ui/widgets/ModFilterWidget.cpp]: Scrolling should stay for the regular combobox, it's just really confusing when it selects multiple items

Comment (Trial97) [reply launcher/ui/widgets/ModFilterWidget.cpp]: ok moved that and also implemented the paint event to make it look like a normal QComboBox.

**Comment by Trial97** [comment]
@getchoo this is the last one 
image

**Review by TayouVR — CHANGES_REQUESTED** [review]
I can't say much about the code at this moment, but functionality:

Everything I have tested seems to work. I like the UI positioning right now, sidebar makes sense and feels natural compared to other places where filters like that exist.

However especially on small screens the three-pane-view + sidebar can be a bit difficult as there is little space left for each panel. Would it be possible to make the three columns' width adjustable with little grab handles? I assume Qt must have something for that.

Otherwise I think this was already suggested, but the providers list takes quite a lot of space up for just two entries... maybe there could be some other way to switch between them. top aligned tabs (like was suggested in discord I think, not sure about here) may work, but not sure how that would look. Probably best to leave it as it is for now?

I had a small UI focus issue with the version dropdown on window, it wouldn't unfocus it when I clicked off it. (needed to use ESC key) Not sure what caused that, but it is consistent, even across themes. This is on windows 10 with the latest CI build. (PrismLauncher-Windows-MSVC-559a811-Debug, 2024-05-18)

Another thing I'm noticing, but that may be API dependent; Modrinth seems to use the tags/categories, etc. additively, showing all results containing any of the tags, as far as I can tell, while curseforge seems to do the opposite. I select e.g. "Server Utility" and "Twitch Integration", it seems to show only a single mod.
Since this is probably just a discrepancy in how the platforms handle search I am not sure we can do much about it, but its something I noticed. If there is some way to define one or the other behaviour for either platform it would be nice to have an option to switch between them.

One more thing with UI/UX; I feel like the filter panel itself should have some sort of "Apply Filters" or just "Apply" button. While I understood pretty quickly that I need to hit search again I think some users may get confused without it.
Or one alternative would be to automatically refresh the search just like while typing the search string. That would probably be even better UX wise, as long as we don't end up over stressing their search API and causing them problems.

**Comment by Trial97** [comment]
> However especially on small screens the three-pane-view + sidebar can be a bit difficult as there is little space left for each panel. Would it be possible to make the three columns' width adjustable with little grab handles? I assume Qt must have something for that.

I will have a look on how to make the columns resizeable
 
> Otherwise I think this was already suggested, but the providers list takes quite a lot of space up for just two entries... maybe there could be some other way to switch between them. top aligned tabs (like was suggested in discord I think, not sure about here) may work, but not sure how that would look. Probably best to leave it as it is for now?

I will give a shot to the resize column that you mentioned.

> I had a small UI focus issue with the version dropdown on window, it wouldn't unfocus it when I clicked off it. (needed to use ESC key) Not sure what caused that, but it is consistent, even across themes. This is on windows 10 with the latest CI build. (PrismLauncher-Windows-MSVC-559a811-Debug, 2024-05-18)

That may not be related to this PR but I will see what I can do(maybe a separate PR)

> Another thing I'm noticing, but that may be API dependent; Modrinth seems to use the tags/categories, etc. additively, showing all results containing any of the tags, as far as I can tell, while curseforge seems to do the opposite. I select e.g. "Server Utility" and "Twitch Integration", it seems to show only a single mod. Since this is probably just a discrepancy in how the platforms handle search I am not sure we can do much about it, but its something I noticed. If there is some way to define one or the other behaviour for either platform it would be nice to have an option to switch between them.

That is clearly API fault. There should be a issue on Prism related to the fact that Modirinth displays fabric mods on forge packs(due to them having support for both) 

> One more thing with UI/UX; I feel like the filter panel itself should have some sort of "Apply Filters" or just "Apply" button. While I understood pretty quickly that I need to hit search again I think some users may get confused without it. Or one alternative would be to automatically refresh the search just like while typing the search string. That would probably be even better UX wise, as long as we don't end up over stressing their search API and causing them problems.


I will test both of them and see what I can fix.

So summary:
- make the panels resizable (to fix the spacing issues)
- add a button to apply  the filters(or apply them on change)
- look into version combobox remaining selected

File: launcher/ui/pages/modplatform/ResourcePage.ui [hunk 3]
```diff
@@ -11,35 +11,55 @@
...
+   </item>
+   <item row="0" column="1">
+    <widget class="QPushButton" name="resourceFilterButton">
+     <property name="text">
+      <string>Filter options</string>
+     </property>
+    </widget>
+   </item>
+   <item row="4" column="0" colspan="4">
+    <widget class="QSplitter" name="spliter">
```
Comment (TayouVR) [inline launcher/ui/pages/modplatform/ResourcePage.ui]:
```suggestion
    <widget class="QSplitter" name="splitter">
```
typo

File: launcher/ui/pages/modplatform/ModPage.cpp [hunk 4]
```diff
@@ -67,17 +68,17 @@ void ModPage::setFilterWidget(unique_qobject_ptr<ModFilterWidget>& widget)
     if (m_filter_widget)
         disconnect(m_filter_widget.get(), nullptr, nullptr, nullptr);
 
+    m_ui->spliter->replaceWidget(0, widget.get());
```
Comment (TayouVR) [inline launcher/ui/pages/modplatform/ModPage.cpp]:
```suggestion
    m_ui->splitter->replaceWidget(0, widget.get());
```
typo

File: launcher/ui/pages/modplatform/ResourcePage.ui [hunk 5]
```diff
@@ -11,35 +11,55 @@
...
+     </property>
+    </widget>
+   </item>
+   <item row="2" column="0" colspan="4">
+    <widget class="ProgressWidget" name="fetchProgress" native="true"/>
+   </item>
+   <item row="0" column="1">
+    <widget class="QPushButton" name="resourceFilterButton">
+     <property name="text">
+      <string>Filter options</string>
```
Comment (TayouVR) [inline launcher/ui/pages/modplatform/ResourcePage.ui]: could this button be on the left of the search bar so it is closer to the filter panel maybe?

**Review by TayouVR — APPROVED** [review]
I think this is pretty good now, version dropdown focus issue aside

File: launcher/ui/pages/modplatform/ResourcePage.ui (hunk 5 above)
Comment (Trial97) [reply launcher/ui/pages/modplatform/ResourcePage.ui]: Yes I can

**Comment by Trial97** [comment]
Should I remove the Search button on the Resource Download dialog?

**Comment by TayouVR** [comment]
> Should I remove the Search button on the Resource Download dialog?

it may still be useful for accessibility, I think I'd keep it around.

**Comment by TayouVR** [comment]
combo box focus issue seems fixed btw with the latest commits.
Also check https://github.com/Trial97/PrismLauncher/pull/3 @Trial97 (idk how to commit directly to the branch from my local git xD

File: launcher/modplatform/modrinth/ModrinthAPI.cpp [hunk 6]
```diff
@@ -120,3 +120,41 @@ QList<ResourceAPI::SortingMethod> ModrinthAPI::getSortingMethods() const
...
+        qWarning() << *response;
+        return categories;
+    }
+
+    try {
+        auto arr = Json::requireArray(doc);
+
+        for (auto val : arr) {
+            auto cat = Json::requireObject(val);
+            auto name = Json::requireString(cat, "name");
```
Comment (getchoo) [inline launcher/modplatform/modrinth/ModrinthAPI.cpp:149]: these names should be converted to title case (or at least the first letter should be uppercase) to look better in the ui. modrinth's api returns them all in lowercase

**Review by getchoo — CHANGES_REQUESTED** [review]
otherwise LGTM and works well. thanks!

File: launcher/modplatform/modrinth/ModrinthAPI.cpp (hunk 6 above)
Comment (Trial97) [reply launcher/modplatform/modrinth/ModrinthAPI.cpp:149]: Done

**Review by Ryex — APPROVED** [review]
Love it, the curseforge API stuff looks correct form what I remember of the changes that allowed searching multiple categories at once and more importantly it works :P

**Comment by Trial97** [comment]
@getchoo can you review this?

**Review by getchoo — CHANGES_REQUESTED** [review]
on opening the options for any instance, the launcher immediately crashes with this log
```
     3.527 D | Component list save performed now for "1.8.9-forge"
     3.531 D | Loading components
     3.531 D | Index is already loaded
     3.531 D | Component list save is scheduled for "1.8.9-forge"
     3.531 D | Resolving dependencies
     3.531 D | "Req: net.minecraft == 1.8.9" Is met.
     3.531 D | "Req: org.lwjgl" Is met.
     3.531 D | Component list update/resolve task succeeded for "1.8.9-forge"
     3.533 D | QProxyStyle NoBigComboBox created for "" QStyleSheetStyle(0x1e60d5704e0)
     3.566 F | ASSERT failure in QList::at: "index out of range", file D:\a\PrismLauncher\Qt\6.7.0\msvc2019_64\include\QtCore/qlist.h, line 434
```

i don't think this was introduced here, but to be safe this should be rebased onto latest `develop` where this problem doesn't occur

**Comment by Trial97** [comment]
> i don't think this was introduced here, but to be safe this should be rebased onto latest `develop` where this problem doesn't occur

I update it .




> on opening the options for any instance, the launcher immediately crashes with this log


~do you mean right click an instance?~ found it is a crash due to a merge I presume

**Review by getchoo — APPROVED** [review]
that was the fix; works great now. thanks!

**Comment by pizzav-xyz** [comment]
when will a version with this added be published?

**Comment by Trial97** [comment]
in 9.0

**Comment by pizzav-xyz** [comment]
> in 9.0

Why not 8.4?

**Comment by Trial97** [comment]
Because minor releases are only for bug fixes or small features

**Pull request merged**